File: assn3/assn3/LZW.cpp
#include <bits/stdc++.h>
using namespace std;
// --- added helpers ---
static vector<string> read_words_from_file(const string& path) {
    ifstream fin(path);
    if (!fin) throw runtime_error("Could not open input file: " + path);
    vector<string> words;
    words.reserve(1 << 20);
    string w;
    while (fin >> w) words.push_back(w);
    if (words.empty()) throw runtime_error("File contained no words.");
    return words;
}

static string join_sampled_words(const vector<string>& words,
                                long long n_words,
                                unsigned long long seed) {
    if (n_words <= 0) throw runtime_error("n_words must be positive.");
    long long use = min<long long>(n_words, (long long)words.size());
    vector<size_t> idx(words.size());
    iota(idx.begin(), idx.end(), 0);
    std::mt19937_64 rng(seed);
    shuffle(idx.begin(), idx.end(), rng);
    vector<string> sample;
    sample.reserve((size_t)use);
    for (long long i = 0; i < use; ++i) sample.push_back(words[idx[(size_t)i]]);
    // join with spaces
    size_t total = 0;
    for (auto& s : sample) total += s.size() + 1;
    string out; out.reserve(total ? total - 1 : 0);
    for (size_t i = 0; i < sample.size(); ++i) {
        if (i) out.push_back(' ');
        out += sample[i];
    }
    return out;
}

vector<int> encoding(string s1)
{
    cout << "Encoding\n";
    unordered_map<string, int> table;
    for (int i = 0; i <= 255; i++) {
        string ch = "";
        ch += char(i);
        table[ch] = i;
    }

    string p = "", c = "";
    p += s1[0];
    int code = 256;
    vector<int> output_code;
    // cout << "String\tOutput_Code\tAddition\n";
    for (int i = 0; i < s1.length(); i++) {
        if (i != s1.length() - 1)
            c += s1[i + 1];
        if (table.find(p + c) != table.end()) {
            p = p + c;
        }
        else {
            // cout << p << "\t" << table[p] << "\t\t" 
                 // << p + c << "\t" << code << endl;
            output_code.push_back(table[p]);
            table[p + c] = code;
            code++;
            p = c;
        }
        c = "";
    }
    // cout << p << "\t" << table[p] << endl;
    output_code.push_back(table[p]);
    return output_code;
}

void decoding(vector<int> op)
{
    cout << "\nDecoding\n";
    unordered_map<int, string> table;
    for (int i = 0; i <= 255; i++) {
        string ch = "";
        ch += char(i);
        table[i] = ch;
    }
    int old = op[0], n;
    string s = table[old];
    string c = "";
    c += s[0];
    cout << s;
    int count = 256;
    for (int i = 0; i < op.size() - 1; i++) {
        n = op[i + 1];
        if (table.find(n) == table.end()) {
            s = table[old];
            s = s + c;
        }
        else {
            s = table[n];
        }
        cout << s;
        c = "";
        c += s[0];
        table[count] = table[old] + c;
        count++;
        old = n;
    }
}
int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    if (argc < 3) {
        cerr << "Usage: " << argv[0] << " <path/to/input.txt> <n_words> [seed]\n";
        return 1;
    }
    const string path = argv[1];
    long long n_words = atoll(argv[2]);
    unsigned long long seed;
    if (argc >= 4) {
        seed = strtoull(argv[3], nullptr, 10);
    } else {
        seed = (unsigned long long)
            chrono::high_resolution_clock::now().time_since_epoch().count();
    }

    try {
        // A) read file
        auto words = read_words_from_file(path);
        // B) choose n words at random (without replacement) and join
        string input = join_sampled_words(words, n_words, seed);
        // C) time compression
        auto t0 = chrono::high_resolution_clock::now();
        vector<int> output_code = encoding(input);
        auto t1 = chrono::high_resolution_clock::now();
        auto micros = chrono::duration_cast<chrono::microseconds>(t1 - t0).count();

        cout << "\n--- Stats ---\n";
        cout << "Selected words: " << min<long long>(n_words, (long long)words.size()) << "\n";
        cout << "Input bytes:    " << input.size() << "\n";
        cout << "Codes emitted:  " << output_code.size() << "\n";
        cout << "Compress time:  " << micros << " microseconds\n";

        // Optional: verify decoding correctness on demand (commented out)
        // decoding(output_code);  // beware: prints the reconstructed text to stdout

    } catch (const exception& e) {
        cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}

File: assn3/assn3/LZW.py
# lzw_benchmark.py
import argparse
import random
import time
from typing import List, Dict

def read_words_from_file(path: str) -> List[str]:
    try:
        with open(path, "r", encoding="utf-8", errors="ignore") as f:
            # whitespace-delimited tokenization
            return f.read().split()
    except OSError as e:
        raise SystemExit(f"Error: could not open file: {path}\n{e}")

def join_sampled_words(words: List[str], n_words: int, seed: int) -> str:
    if n_words <= 0:
        raise SystemExit("Error: n_words must be positive.")
    use = min(n_words, len(words))
    idx = list(range(len(words)))
    rng = random.Random(seed)
    rng.shuffle(idx)
    sample = [words[i] for i in idx[:use]]
    return " ".join(sample)

def encoding(s: str) -> List[int]:
    """LZW encoding (string -> list of integer codes)."""
    if not s:
        return []
    # initialize dictionary with single-byte strings
    table: Dict[str, int] = {chr(i): i for i in range(256)}

    p = s[0]
    code = 256
    out: List[int] = []

    # Note: matches your C++ control flow (p + next char probe)
    for i in range(len(s)):
        c = s[i + 1] if i + 1 < len(s) else ""
        pc = p + c
        if pc in table:
            p = pc
        else:
            out.append(table[p])
            table[pc] = code
            code += 1
            p = c

    # flush last token
    out.append(table[p])
    return out

def decoding(op: List[int]) -> str:
    """LZW decoding (list of integer codes -> string)."""
    if not op:
        return ""
    table: Dict[int, str] = {i: chr(i) for i in range(256)}

    old = op[0]
    s = table[old]
    c = s[0]
    count = 256
    out = [s]

    for i in range(len(op) - 1):
        n = op[i + 1]
        if n not in table:
            s = table[old] + c
        else:
            s = table[n]
        out.append(s)
        c = s[0]
        table[count] = table[old] + c
        count += 1
        old = n

    return "".join(out)

def main():
    parser = argparse.ArgumentParser(
        description="Read words from file, sample N at random, LZW-compress, and time it."
    )
    parser.add_argument("path", help="Path to input text file")
    parser.add_argument("n_words", type=int, help="Number of words to sample")
    parser.add_argument(
        "seed",
        nargs="?",
        type=int,
        help="Optional RNG seed (int). If omitted, current time (ns) is used.",
    )
    args = parser.parse_args()

    # If no seed provided, use current time in ns (like C++ high-res clock)
    seed = args.seed if args.seed is not None else time.time_ns()

    words = read_words_from_file(args.path)
    if not words:
        raise SystemExit("Error: file contains no words.")

    # Build input text by sampling words without replacement
    input_text = join_sampled_words(words, args.n_words, seed)

    # Time compression
    t0 = time.perf_counter_ns()
    codes = encoding(input_text)
    t1 = time.perf_counter_ns()
    micros = (t1 - t0) // 1_000  # microseconds

    # Report stats (mirroring C++ style)
    print(f"Selected words: {min(args.n_words, len(words))}")
    print(f"Input bytes:    {len(input_text)}")
    print(f"Output codes:   {len(codes)} (LZW variable-width codes)")
    print(f"Compress time:  {micros} microseconds")
    print(f"Seed used:      {seed}")

    # # Write codes (space-separated) for inspection (not a compact binary format)
    # with open("compressed_codes.txt", "w", encoding="utf-8") as fout:
    #     fout.write(" ".join(str(x) for x in codes))
    # print("Wrote codes to compressed_codes.txt")


if __name__ == "__main__":
    main()

